index.js

```diff
@@ -4,43 +4,370 @@ var MongoClient = require('mongodb').MongoClient;
+                callback(false, obj);
+            });
+        }];
+
+        var resultArray = [];
+
+        async.parallel(requests, function (err, results) {
+            if (err) {
+                console.log(err);
+                res.send(500, "Server Error");
+                return;
+            }
+
+            var query = results[0].value;
+
+            for (var i = 0; i < userCart.length; i++) {
+                var queriedItem = query.find(function(item) {
+                    return item._id === userCart[i].itemId;
+                });
+
+                if(queriedItem) {
+                    resultArray.push({
+                        item: queriedItem,
+                        quantity: userCart[i].quantity,
+                        imageUrl: userCart[i].imageUrl
+                    });
+                    subTotal += userCart[i].quantity * queriedItem.price;
+                }
+            }
+
+            res.send({cart: resultArray, subTotal: subTotal});
+        });
+    } else {
+        res.send({cart: [], subTotal: 0});
+    }
+}
+
+app.get('/cart', function(req, res) {
+    var cartId = req.query.cartId;
+
+    if(cartId === undefined){
+        var cartId = _.uniqueId('cartId-');
+        cartModel.sessions.push({cart: [], cartId: cartId });
+        res.send({cartId: cartId});
+    } else {
+        var userCart = cartModel.cartById(cartId);
+
+        if(userCart === undefined){
+            cartModel.sessions.push({cartId: cartId, cart: []});
+        }
+    }
+    sendCartWithSubTotal(cartId, res);
+});
+
+var nodeMailer = new nodemailer();
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({extended: false}));
+app.post("/sendMail", function(req,res) {
+   var email = req.body.userInfo.email;
+   var products = req.body.products;
+   var sendable = "";
+
+   for (var i=0;i<products.length;i++) {
+       sendable += products[i].item.title + "\t" + products[i].quantity + "\t" + products[i].item.price * products[i].quantity + "\n";
+   }
+   res.send(nodeMailer.sendMail(email, 'Epam-grocery-webshop order', sendable));
+});
+
+app.post("/login", function(req, res) {
+    var name = req.body.name;
+    var password = req.body.password;
+    var url = BASEAPIURL + "/users?$filter=name eq '" + name + "' and password eq '" + md5(password) + "'";
+
+    request(url, function (err, response, body) {
+        // JSON body
+        if (err) {
+            console.log(err);
+            callback(true);
+            return;
+        }
+        obj = JSON.parse(body);
+
+        if(!!obj.value[0]){
+            var tokenArrayElement = tokenArray.find(function(element) {
+                return element.name === name;
+            });
+            
+            if(!!tokenArrayElement){
+                res.send({token: tokenArrayElement.token});
+            } else {
+                var token = tokenGenerator.generate();
+                tokenArray.push({name: name, token: token, id: obj.value[0]._id});
```

Comment:
if you use an object instead of tokenArray, you could save the token as key. In this case you could find the token - user pairs easily.